lib/edit-site-export.php

```diff
@@ -0,0 +1,64 @@
+<?php
+/**
+ * REST endpoint for exporting the contents of the Edit Site Page editor.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Output a ZIP file with an export of the current templates
+ * and template parts from the site editor, and close the connection.
+ */
+function gutenberg_edit_site_export() {
+	// Create ZIP file and directories.
+	$filename = tempnam( get_temp_dir(), 'edit-site-export' );
+	$zip      = new ZipArchive();
+	$zip->open( $filename, ZipArchive::OVERWRITE );
+	$zip->addEmptyDir( 'theme' );
+	$zip->addEmptyDir( 'theme/block-templates' );
+	$zip->addEmptyDir( 'theme/block-template-parts' );
+
+	// Load files into ZIP file.
+	foreach ( get_template_types() as $template_type ) {
+		// Skip 'embed' for now because it is not a regular template type.
+		// Skip 'index' because it's a fallback that we handle differently.
+		if ( in_array( $template_type, array( 'embed', 'index' ), true ) ) {
+			continue;
+		}
+
+		$current_template = gutenberg_find_template_post_and_parts( $template_type );
+		if ( isset( $current_template ) ) {
+			$zip->addFromString( 'theme/block-templates/' . $current_template['template_post']->post_name . '.html', $current_template['template_post']->post_content );
+
+			foreach ( $current_template['template_part_ids'] as $template_part_id ) {
+				$template_part = get_post( $template_part_id );
+				$zip->addFromString( 'theme/block-template-parts/' . $template_part->post_name . '.html', $template_part->post_content );
+			}
+		}
+	}
+
+	// Send back the ZIP file.
+	$zip->close();
+	header( 'Content-Type: application/zip' );
+	header( 'Content-Disposition: attachment; filename=edit-site-export.zip' );
+	header( 'Content-Length: ' . filesize( $filename ) );
+	flush();
+	echo readfile( $filename );
```

Comment:
Shouldn't we clean up, delete the file? Especially since it's using temporary filenames, so we're not just overwriting the same file.

Reply:
Is there an issue with keeping it around? Deleting it would technically be more work.

Reply:
If we do make changes here we should also switch to `wp_tempnam` which I missed the first time round.

Reply:
Curious if that's just for consistency/style or is there some benefit to it?

Reply:
I think its mostly consistency and style, but it does appear to have some edge case handling.

Reply:
> Is there an issue with keeping it around?

I don't think we can make assumptions on how often the location returned by `get_temp_dir()` is flushed. We must assume that ZIP files would proliferate over time.

> Deleting it would technically be more work.

How so?

Reply:
>How so?

You always have to create it/overwrite it. Deleting it is another set of operations we could avoid, although I doubt it makes a measurable difference.

I think we can be safe here by using the same file name every time. That way it just overwrites the same file if it hasn't been cleaned up already.

Reply:
For the record, @epiqueras and I talked about this over Slack, and my recommendation was to ensure good random filenames (WordPress's `wp_tempnam` might be better at avoiding naming conflicts) and ensure we delete the file in the moment.

This stands in contrast with adopting a fixed filename so as to overwrite the file over time. This might be enough in simple sites, but is bound to cause race-condition-type problems in any larger installation, and especially in multisite or highly concurrent scenarios.